crates/ide_assists/src/handlers/move_guard.rs

```diff
@@ -115,73 +116,119 @@ pub(crate) fn move_arm_cond_to_match_guard(acc: &mut Assists, ctx: &AssistContex
+                        let to_insert = block.dedent(dedent.into()).syntax().text();
+                        edit.insert(then_arm_end, to_insert)
+                    }
+                }
+            }
+            match tail {
+                Some(Tail::IfLet(e)) => {
+                    cov_mark::hit!(move_guard_ifelse_iflet_tail);
+                    let guard = format!("\n{}{} => ", spaces, match_pat);
+                    // Put the if-let expression in a block
+                    let iflet_expr: Expr = e.reset_indent().indent(1.into()).into();
+                    let iflet_block =
+                        make::block_expr(std::iter::empty(), Some(iflet_expr)).indent(indent_level);
+                    edit.insert(then_arm_end, guard);
+                    edit.insert(then_arm_end, iflet_block.syntax().text());
+                }
+                Some(Tail::Else(e)) => {
+                    cov_mark::hit!(move_guard_ifelse_else_tail);
+                    let guard = format!("\n{}{} => ", spaces, match_pat);
+                    edit.insert(then_arm_end, guard);
+                    let only_expr = e.statements().next().is_none();
+                    match &e.tail_expr() {
+                        Some(expr) if only_expr => {
+                            cov_mark::hit!(move_guard_ifelse_expr_only);
+                            edit.insert(then_arm_end, expr.syntax().text());
+                            edit.insert(then_arm_end, ",");
+                        }
+                        _ => {
+                            let to_insert = e.dedent(dedent.into()).syntax().text();
+                            edit.insert(then_arm_end, to_insert)
+                        }
                     }
                 }
+                _ => {
+                    cov_mark::hit!(move_guard_ifelse_notail);
+                }
             }
         },
     )
 }
 
+#[derive(Debug)]
+enum Tail {
+    Else(BlockExpr),
+    IfLet(IfExpr),
+}
+
+// Parses an if-else-if chain to get the conditons and the then branches until we encounter an else
+// branch, an if-let branch or the end.
+fn parse_if_chain(if_expr: IfExpr) -> Option<(Vec<(Condition, BlockExpr)>, Option<Tail>)> {
+    let mut conds_blocks = Vec::new();
+    let mut curr_if = if_expr;
+    let mut applicable = false;
+    let tail: Option<Tail> = loop {
+        let cond = curr_if.condition()?;
+        if cond.is_pattern_cond() {
+            break Some(Tail::IfLet(curr_if));
+        }
+        conds_blocks.push((cond, curr_if.then_branch()?));
+        applicable = true;
+        match curr_if.else_branch() {
+            Some(ElseBranch::IfExpr(e)) => {
+                curr_if = e;
+            }
+            Some(ElseBranch::Block(b)) => {
+                break Some(Tail::Else(b));
+            }
+            None => break None,
+        }
+    };
+    if !applicable {
+        // The first if branch is an if-let branch
+        return None;
+    }
```

Comment:
Shouldn't this fail for all if-let branches in general and not just the first one?

Reply:
If the first if branch is not if-let, we can convert part of the if else chain to the match arms and leave the if-let branch in an arm without a guard. but I think not allowing if-let branches at all is also ok.